src/install/isolated_install/Installer.zig

```diff
@@ -1284,26 +1381,60 @@ pub const Installer = struct {
-                .string_buf = string_buf,
-                .extern_string_buf = extern_string_buf,
-                .seen = &seen,
-                .node_modules_path = &node_modules_path,
-                .abs_target_buf = link_target_buf,
-                .abs_dest_buf = link_dest_buf,
-                .rel_buf = link_rel_buf,
-            };
+            var target_node_modules_path: ?bun.AbsPath(.{}) = null;
+            defer if (target_node_modules_path) |*path| path.deinit();
+            const package_name = strings.StringOrTinyString.init(alias.slice(string_buf));
+
+            const target_package_name = this.maybeReplaceNodeModulesPath(
+                entry_node_ids,
+                node_pkg_ids,
+                pkg_name_hashes,
+                pkg_resolutions_lists,
+                pkg_resolutions_buffer,
+                pkg_metas,
+                pkg_id,
+                &target_node_modules_path,
+            ) orelse package_name;
+
+            var can_retry_without_native_binlink_optimization = target_node_modules_path != null;
+            while (true) {
+                var bin_linker: Bin.Linker = .{
+                    .bin = bin,
+                    .global_bin_path = this.manager.options.bin_path,
+                    .package_name = package_name,
+                    .string_buf = string_buf,
+                    .extern_string_buf = extern_string_buf,
+                    .seen = &seen,
+                    .node_modules_path = &node_modules_path,
+                    .target_node_modules_path = if (target_node_modules_path) |*path| path else &node_modules_path,
+                    .target_package_name = if (target_node_modules_path != null) target_package_name else package_name,
+                    .abs_target_buf = link_target_buf,
+                    .abs_dest_buf = link_dest_buf,
+                    .rel_buf = link_rel_buf,
+                };
 
-            bin_linker.link(false);
+                bin_linker.link(false);
+
+                if (can_retry_without_native_binlink_optimization and (bin_linker.skipped_due_to_missing_bin or bin_linker.err != null)) {
+                    can_retry_without_native_binlink_optimization = false;
+                    target_node_modules_path.?.deinit();
+                    target_node_modules_path = null;
+                    if (PackageManager.verbose_install) {
+                        Output.prettyErrorln("<d>[Bin Linker]<r> {s} -> {s} retrying without native bin link", .{
+                            package_name.slice(),
+                            target_package_name.slice(),
+                        });
+                    }
+                    continue;
+                }
 
-            if (bin_linker.err) |err| {
-                return err;
+                if (bin_linker.err) |err| {
+                    return err;
+                }
+
+                break;
```

Comment:
this should be more simple. i'd delete `can_retry_without_native_binlink_optimization` and the `while(true)`. after the first link do the second link:
```
if (target_node_modules_path != null and (bin_linker.skipped_due_to_missing_bin or bin_linker.err != null)) {
    target_node_modules_path.?.deinit();
    target_node_modules_path = null;

    bin_linker.target_node_modules_path = &node_modules_path;
    bin_linker.target_package_name = package_name;

    if (PackageManager.verbose_install) {
        Output.prettyErrorln("<d>[Bin Linker]<r> {s} -> {s} retrying without native bin link", .{
            package_name.slice(),
            target_package_name.slice(),
        });
    }

    bin_linker.link(false);
}
```